jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/WebSocketNegotiationTest.java

```diff
@@ -353,4 +359,81 @@ public void testInvalidUpgradeRequestNoKey() throws Exception
+            Arguments.of("ext1, @int1, ext2", "ext1, @int1, ext2"),
+            Arguments.of("ext1, ext2, @int1", "ext1, ext2, @int1"),
+            Arguments.of("@int1, ext1, @int2, ext2, @int3", "@int1, ext1, @int2, ext2, @int3"),
+            Arguments.of("ext1, ext1, ext1, @int1, ext2", "ext1, @int1, ext2"),
+            Arguments.of("ext1, @int1, ext1, ext1, ext2", "ext1, @int1, ext2"),
+            Arguments.of("ext1, ext2, ext3, @int1", "ext1, ext2, ext3, @int1")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("internalExtensionScenarios")
+    public void testClientRequestedInternalExtensions(String reqExts, String negExts) throws Exception
+    {
+        // Add some simple Extensions for to make test examples clearer.
+        WebSocketExtensionRegistry extRegistry = components.getExtensionRegistry();
+        extRegistry.register("ext1", EmptyExtension.class);
+        extRegistry.register("ext2", EmptyExtension.class);
+        extRegistry.register("ext3", EmptyExtension.class);
+        extRegistry.register("@int1", EmptyExtension.class);
+        extRegistry.register("@int2", EmptyExtension.class);
+        extRegistry.register("@int3", EmptyExtension.class);
+
+        TestFrameHandler clientHandler = new TestFrameHandler();
+        CompletableFuture<String> extensionHeader = new CompletableFuture<>();
+        ClientUpgradeRequest upgradeRequest = ClientUpgradeRequest.from(client, server.getUri(), clientHandler);
+        upgradeRequest.setSubProtocols("test");
+        upgradeRequest.addExtensions(reqExts.split(","));
+        upgradeRequest.addListener(new UpgradeListener()
+        {
+            @Override
+            public void onHandshakeResponse(HttpRequest request, HttpResponse response)
+            {
+                extensionHeader.complete(response.getHeaders().get(HttpHeader.SEC_WEBSOCKET_EXTENSIONS));
+            }
+        });
+
+        // Connect to the client then close the Session.
+        client.connect(upgradeRequest).get(5, TimeUnit.SECONDS);
+        clientHandler.sendClose();
+        assertTrue(clientHandler.closed.await(5, TimeUnit.SECONDS));
+        assertNull(clientHandler.getError());
+
+        // We had no internal Extensions in the response headers.
+        assertThat(extensionHeader.get(5, TimeUnit.SECONDS), not(containsString("@")));
+
+        // The list of Extensions on the client contains the internal Extensions.
+        StringBuilder negotiatedExtensions = new StringBuilder();
+        List<Extension> extensions = ((WebSocketCoreSession)clientHandler.coreSession).getExtensionStack().getExtensions();
+        for (int i = 0; i < extensions.size(); i++)
+        {
+            negotiatedExtensions.append(extensions.get(i).getConfig().getParameterizedName());
+            if (i != extensions.size() - 1)
+                negotiatedExtensions.append(", ");
+        }
+        assertThat(negotiatedExtensions.toString(), is(negExts));
+    }
```

Comment:
This test is fine, but you want to test also the edge cases: just one extension, zero extensions, one internal and one external, etc. - you are always testing for more than 1 extension.